inc/api/api.class.php

```diff
@@ -2891,4 +2894,234 @@ public function handleDepreciation(string $itemtype): string {
+      if (!isset($actions[$action_key])) {
+         return $this->returnError(
+            "Invalid action key parameter, run 'getMassiveActions' endpoint to see available keys",
+            400,
+            "ERROR_MASSIVEACTION_KEY"
+         );
+      }
+
+      // Get massive action for the given key
+      $ma = new MassiveAction([
+         'action'     => $action_key,
+         'actions'    => $actions,
+         'items'      => [],
+         'is_deleted' => $is_deleted
+      ], [], 'specialize');
+
+      // Capture form display
+      ob_start();
+      $ma->showSubForm();
+      $html = ob_get_clean();
+
+      // Parse html to find all non hidden inputs, textareas and select
+      $inputs = [];
+      $crawler = new Crawler($html);
+      $crawler->filterXPath('//input')->each(function (Crawler $node, $i) use (&$inputs) {
+         if ($node->attr('type') != "hidden") {
+            $inputs[] = [
+               'name' => $node->attr('name'),
+               'type' => $node->attr('type'),
+            ];
+         }
+      });
+      $crawler->filterXPath('//select')->each(function (Crawler $node, $i) use (&$inputs) {
+         $type = 'select';
+         if (Toolbox::startsWith($node->attr('id'), 'dropdown_')) {
+            $type = 'dropdown';
+         }
+         $inputs[] = [
+            'name' => $node->attr('name'),
+            'type' => $type,
+         ];
+      });
+      $crawler->filterXPath('//textarea')->each(function (Crawler $node, $i) use (&$inputs) {
+         $inputs[] = [
+            'name' => $node->attr('name'),
+            'type' => 'text',
+         ];
+      });
+
+      return $this->returnResponse($inputs);
+   }
+
+
+   /**
+    * "applyMassiveAction" endpoint.
+    * Execute the given massive action
+    *
+    * @param string        $itemtype      Target itemtype
+    * @param string|null   $action_key    Target massive action
+    * @param array         $ids           Ids of items to execute the action on
+    * @param array         $params        Action parameters
+    * @return array
+    */
+   public function applyMassiveAction(
+      string $itemtype,
+      ?string $action_key,
+      array $ids,
+      array $params
+   ) {
+      if (is_null($action_key)) {
+         return $this->returnError(
+            "Missing action key, run 'getMassiveActions' endpoint to see available keys",
+            400,
+            "ERROR_MASSIVEACTION_KEY"
+         );
+      }
+
+      // Get processor
+      $action = explode(':', $action_key);
+      $processor = $action[0];
+
+      // Guess "is_deleted" from an last item
+      $item = new $itemtype();
+      if (!$item->getFromDB(current($ids))) {
+         return $this->returnError(
+            "Failed to load item (itemtype = '$itemtype', id = '" . current($ids) . "')",
+            400,
+            "ERROR_ITEM_NOT_FOUND"
+         );
+      }
+      $is_deleted = $item->isDeleted();
```

Comment:
Are you sure the `is_deleted` is usefull here ? As far as I see, it is only used in the `initial` stage of the `MassiveAction` constructor and in the `MassiveAction::getAllMassiveActions()` method.

If it is mandatory, maybe it is better to require it in the API method input, in order to be sure that developpers will understand that calling a massive action on both deleted and not deleted items at the same time is not a good idea.
Another solution could be to guess this flag based on requested action. Indeed, you could call `MassiveAction::getAllMassiveActions()` with both `is_deleted=0` and `is_deleted=1` and check if the requested action is in the first or second list.

Reply:
is_deleted isn't used for this step, better to remove it -> 908543ecf4.